powersimdata/data_access/launcher.py

```diff
@@ -0,0 +1,207 @@
+
+
+class SSHLauncher(Launcher):
+    def _run_script(self, script, extra_args=None):
+        """Returns running process
+
+        :param str script: script to be used.
+        :param list extra_args: list of strings to be passed after scenario id.
+        :return: (*subprocess.Popen*) -- process used to run script
+        """
+        if extra_args is None:
+            extra_args = []
+
+        engine = self.scenario._scenario_info["engine"]
+        path_to_package = posixpath.join(server_setup.MODEL_DIR, engine)
+        folder = "pyreise" if engine == "REISE" else "pyreisejl"
+
+        path_to_script = posixpath.join(path_to_package, folder, "utility", script)
+        cmd_pythonpath = [f'export PYTHONPATH="{path_to_package}:$PYTHONPATH";']
+        cmd_pythoncall = [
+            "nohup",
+            "python3",
+            "-u",
+            path_to_script,
+            self.scenario.scenario_id,
+        ]
+        cmd_io_redirect = ["</dev/null >/dev/null 2>&1 &"]
+        cmd = cmd_pythonpath + cmd_pythoncall + extra_args + cmd_io_redirect
+        process = self.scenario._data_access.execute_command_async(cmd)
+        print("PID: %s" % process.pid)
+        return process
+
+    def _launch(self, threads=None, solver=None, extract_data=True):
+        """Launch simulation on server, via ssh.
+
+        :param int/None threads: the number of threads to be used. This defaults to None,
+            where None means auto.
+        :param str solver: the solver used for optimization. This defaults to
+            None, which translates to gurobi
+        :param bool extract_data: whether the results of the simulation engine should
+            automatically extracted after the simulation has run. This defaults to True.
+        :raises TypeError: if extract_data is not a boolean
+        :return: (*subprocess.Popen*) -- new process used to launch simulation.
+        """
+        extra_args = []
+        if threads is not None:
+            # Use the -t flag as defined in call.py in REISE.jl
+            extra_args.append("--threads " + str(threads))
+
+        if solver is not None:
+            extra_args.append("--solver " + solver)
+
+        if not isinstance(extract_data, bool):
+            raise TypeError("extract_data must be a bool")
+        if extract_data:
+            extra_args.append("--extract-data")
+
+        return self._run_script("call.py", extra_args=extra_args)
+
+    def extract_simulation_output(self):
+        """Extracts simulation outputs {PG, PF, LMP, CONGU, CONGL} on server.
+
+        :return: (*subprocess.Popen*) -- new process used to extract output
+            data.
+        """
+        print("--> Extracting output data on server")
+        return self._run_script("extract_data.py")
+
+    def check_progress(self):
+        print("Information is available on the server.")
+
+
+class HttpLauncher(Launcher):
+    def _launch(self, threads=None, solver=None, extract_data=True):
```

Comment:
Do we need `extract_data` here, given it is not passed to `request` call anyway.

Reply:
Yeah each `_launch` method has to have the same signature, since we call it the same way (from `Launcher.launch_simulation`) without knowing which subclass it is. I think we might end up passing it to the request eventually though.